File: 061_Deck.rb
# 061 - Deck（★2）
# deque

Q = gets.to_i
TX = Array.new(Q) { gets.split.map(&:to_i) }

# rubyデフォルト配列使用コード
cards, ans = [], []
TX.each do |t, x|
  case t
  when 1
    cards.unshift(x)
  when 2
    cards << x
  when 3
    ans << cards[x - 1]
  end
end

puts ans

# deque作成コード
deque, ans = [], []
cl = cr = 10 ** 5

TX.each do |t, x|
  case t
  when 1
    deque[cl - 1] = x
    cl -= 1
  when 2
    deque[cr] = x
    cr += 1
  when 3
    ans << deque[cl + x - 1]
  end
end

puts ans

File: kotlin/061_Deck.kt
/**
 * 061 - Deck（★2）
 * deque
 */

 fun main() {
  val Q = readLine()!!.toInt()
  val TX = Array(Q) { readLine()!!.split(" ").map(String::toInt) }
  val deque = mutableListOf<Int>()
  val ans = mutableListOf<Int>()

  TX.forEach {(t, x) ->
    when (t) {
      1 -> deque.add(0, x)
      2 -> deque.add(x)
      3 -> ans.add(deque[x - 1])
    }
  }

  ans.forEach(::println)
}
